### ISSUE TITLE
`with` - warning: this clause cannot match because of different types/sizes

### ISSUE THREAD

**Issue by pgrunwald** [elixir-lang/elixir#6380]
### Environment

* Elixir 1.4.4 and Elixir 1.5.0
* Erlang/OTP 20.0
* Operating system: macOS 10.12.5 (16F73) / Darwin 16.6.0 

### Current behavior

Since upgrading to Elixir 1.5.0, I have been receiving warning messages for `with` statements where one of the matches has the right-hand side "wrapped" to help with matching without needing to make a separate function. Consider the following contrived example:

```elixir
defmodule TestModule do
  def test_func do
    with \
      {:first, float1} when is_float(float1) <- {:first, rand_float_or_nil()},
      {:second, float2} when is_float(float2) <- {:second, rand_float_or_nil()}
    do
      {:ok, float1 + float2}
    else
      {:first, nil} -> {:error, "first number is nil"}
      {:second, nil} -> {:error, "second number is nil"}
    end
  end

  defp rand_float_or_nil do
    random_float = :rand.uniform()

    if random_float > 0.5, do: random_float, else: nil
  end
end
```

When compiling with Elixir 1.5.0, a module like the one above results in the following compiler warnings:

```
Compiling 1 file (.ex)
warning: this clause cannot match because of different types/sizes
  lib/test_app/test_module.ex:9

warning: this clause cannot match because of different types/sizes
  lib/test_app/test_module.ex:10
```

However, it's clear that all clauses do in fact match:

```
Erlang/OTP 20 [erts-9.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Interactive Elixir (1.5.0) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> TestModule.test_func()
{:error, "first number is nil"}
iex(2)> TestModule.test_func()
{:ok, 1.6204920530174767}
iex(3)> TestModule.test_func()
{:error, "first number is nil"}
iex(4)> TestModule.test_func()
{:error, "second number is nil"}
```

### Expected behavior

The warning messages do not appear when compiling with Elixir 1.4.4 and the function operates as expected.

```
Erlang/OTP 20 [erts-9.0] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Interactive Elixir (1.4.4) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> TestModule.test_func()
{:error, "first number is nil"}
iex(2)> TestModule.test_func()
{:ok, 1.6464043208507562}
iex(3)> TestModule.test_func()
{:error, "first number is nil"}
iex(4)> TestModule.test_func()
{:error, "first number is nil"}
iex(5)> TestModule.test_func()
{:error, "second number is nil"}
```

**Comment by michalmuskala** [comment]
In a way, this is duplicate of https://github.com/elixir-lang/elixir/issues/6378 and consequence of the chosen approach to making `with` tail recursive.

**Comment by josevalim** [comment]
Thank you @pgrunwald! This has been fixed in master and in the v1.5 branch. 

We should have a new release in a week or two. We just need to grab the first round of regressions.

**Comment by pgrunwald** [comment]
Thanks for the prompt response, @josevalim. Congratulations on the 1.5 release (everyone) and keep up the great work!
